src/docs/arch-hs.dox

```diff
@@ -4,5 +4,111 @@ The History Store in WiredTiger tracks old (all but the latest committed)
+- \c stop timestamp
+- \c durable timestamp
+- \c update type
+- \c value.
+
+Note that every update in the Hisotry Store has a valid stop timestamp and transaction associated
+with it, except for cases where the update preceded a prepared update. The stop timestamp
+is fixed in this case once prepared update is resolved.
+
+@section arch_hs_initialize Historty Store initialization
+WiredTiger checks for an existing History Store table at startup and creates a new table if it is
+not able to find one. History Store table is usually created when WiredTiger creates a new database,
+with one exception where WiredTiger is openning an existing database created by an older version of
+WiredTiger that didn't support Histoty Store feature. WiredTiger uses \c WiredTigerHS.wt as the
+filename for the History Stable table.
+
+@section arch_hs_cursor Historty Store cursor interface
+WiredTiger uses a cursor interface for History Store table to make it easier for different modules
+to perform data operations on the History Strore table. A new cursor can be opened by calling
+\c __wt_curhs_open(). History Store cursor implementation supports most of the standard \c WT_CURSOR
+methods, with one key exception being \c WT_CURSOR::search() method. Recall that the key contains a
+counter value and it is expected that caller can not possibly know the correct counter value to use.
+Therefore, \c WT_CURSOR::search_near() is used for all search operations. Two helper functions,
+\c __wt_curhs_search_near_before() and \c __wt_curhs_search_near_after(), have been provided to
+facilitate searching the required record in the History Store table.
+
+@section arch_hs_cursor_visibilty Historty Store cursor interface and visibility rules
+When using the Historty Store cursor interface, the user can configure the type of visibility checks
+that are performed on the records. The behavhiour is controlled by a set of cursor flags:
+
+- When the flag \c WT_CURSTD_HS_READ_ALL is set on the cursor, no visibility checks are performed
+on the records returned to the API user. This means that cursor interface can even return a record
+with a globally visible tombstone. When this flag is set, it suppresses the effect of other
+visibility flags.
+
+- When the flag \c WT_CURSTD_HS_READ_COMMITTED is set, cursor interface can return any record except
+for records that have a globally visible tombstone. This flag has no effect when the flag
+\c WT_CURSTD_HS_READ_ALL is set.
+
+- When none of the flags discussed above are set on the cursor, snapshot based visibility checks are
+performed on all records returned to the API user. History Store cursor users must use one of
+the flags discussed above if the caller thread is running in a lower isolation level and
+doesn't hold a valid snapshot.
+
+@section arch_hs_reconciliation Historty Store and reconciliation
+When a modified user page is processed by either checkpoint or eviction module, the reconciliation
```

Comment:
I think you can improve this section. "When a dirty page is reconciled on a user file b-tree the update chain is examined and the latest committed update is chosen as the on-disk value. All older updates are added to the history store table assuming they are not yet obsolete. Additionally any out of order timestamps will be corrected." 

Side note: I think the transactions doc describes what out of order timestamps are and should be referenced using doxygen the first time you introduce them in this doc.

Reply:
Unfortunately, I can't find any other reference to out-of-order timestamps in the architecture guide. 

@quchenhao are you aware of any other place where we discuss them?